### PR TITLE
[SPARK-22675] [SQL] Refactoring PropagateTypes in TypeCoercion

### PR DESCRIPTION
**Pull request by gatorsmile**
## What changes were proposed in this pull request?
PropagateTypes are called twice in TypeCoercion. We do not need to call it twice. Instead, we should call it after each change on the types. 

## How was this patch tested?
The existing tests

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by gatorsmile** [comment]
cc @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 1]
```diff
@@ -850,3 +834,45 @@ object TypeCoercion {
     }
   }
 }
+
+trait TypePropagation extends Logging {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: how about
```
trait TypeCoercionRule extends Rule[LogicalPlan] {
  protected def coerciveType(plan: LogicalPlan): LogicalPlan

  def apply(plan: LogicalPlan): LogicalPlan = {
    val newPlan = coerciveType(plan)
    ...
  }
}
```

File: sql/core/src/test/scala/org/apache/spark/sql/TPCDSQuerySuite.scala [hunk 2]
```diff
@@ -39,6 +40,8 @@ class TPCDSQuerySuite extends QueryTest with SharedSQLContext with BeforeAndAfte
    */
   protected override def afterAll(): Unit = {
     try {
+      // For debugging dump some statistics about how much time was spent in various optimizer rules
+      logWarning(RuleExecutor.dumpTimeSpent())
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/TPCDSQuerySuite.scala]: use `logInfo`?

Comment (gatorsmile) [reply sql/core/src/test/scala/org/apache/spark/sql/TPCDSQuerySuite.scala]: Let me change all the other places.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala (hunk 1 above)
Comment (gatorsmile) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: Sure.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DecimalPrecision.scala [hunk 3]
```diff
@@ -58,7 +58,7 @@ import org.apache.spark.sql.types._
  * - FLOAT and DOUBLE cause fixed-length decimals to turn into DOUBLE
  */
 // scalastyle:on
-object DecimalPrecision extends Rule[LogicalPlan] {
+object DecimalPrecision extends Rule[LogicalPlan] with TypePropagation {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DecimalPrecision.scala]: one benefit of naming it `TypeCoercionRule` is, we can just do `object DecimalPrecision extends TypeCoercionRule`

**Comment by gatorsmile** [comment]
BTW, I did not change all the rules in `object TypeCoercion` to propagate the types, because not all of them need to propagate the types.

**Comment by cloud-fan** [comment]
LGTM

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**